packages/firebase_auth/firebase_auth_platform_interface/lib/src/method_channel_firebase_auth.dart

```diff
@@ -0,0 +1,481 @@
+        }
+        break;
+      case 'phoneCodeAutoRetrievalTimeout':
+        final int handle = call.arguments['handle'];
+        final PhoneCodeAutoRetrievalTimeout codeAutoRetrievalTimeout =
+            _phoneAuthCallbacks[handle].codeAutoRetrievalTimeout;
+        final String verificationId = call.arguments['verificationId'];
+        codeAutoRetrievalTimeout(verificationId);
+        break;
+    }
+  }
+
+  void _onAuthStageChangedHandler(MethodCall call) {
+    final Map<dynamic, dynamic> data = call.arguments['user'];
+    final int id = call.arguments['id'];
+
+    final PlatformUser currentUser = data != null ? _decodeUser(data) : null;
+    _authStateChangedControllers[id].add(currentUser);
+  }
+}
+
+PlatformUser _decodeUser(Map<dynamic, dynamic> data) {
+  final List<dynamic> rawProviderData = data['providerData'];
+  final List<Map<dynamic, dynamic>> castProviderData =
+      rawProviderData.cast<Map<dynamic, dynamic>>();
+  final List<PlatformUserInfo> providerData =
+      castProviderData.map<PlatformUserInfo>(_decodeUserInfo).toList();
+  return PlatformUser(
+    providerId: data['providerId'],
+    uid: data['uid'],
+    displayName: data['displayName'],
+    photoUrl: data['photoUrl'],
+    email: data['email'],
+    phoneNumber: data['phoneNumber'],
+    isAnonymous: data['isAnonymous'],
+    isEmailVerified: data['isEmailVerified'],
+    creationTimestamp: data['creationTimestamp'],
+    lastSignInTimestamp: data['lastSignInTimestamp'],
+    providerData: providerData,
+  );
+}
+
+PlatformUserInfo _decodeUserInfo(Map<dynamic, dynamic> data) {
+  return PlatformUserInfo(
+    providerId: data['providerId'],
+    uid: data['uid'],
+    displayName: data['displayName'],
+    photoUrl: data['photoUrl'],
+    email: data['email'],
+    phoneNumber: data['phoneNumber'],
+  );
+}
+
+PlatformAuthResult _decodeAuthResult(Map<dynamic, dynamic> data) {
+  final PlatformUser user = _decodeUser(data['user']);
+  final PlatformAdditionalUserInfo additionalUserInfo =
+      _decodeAdditionalUserInfo(data['additionalUserInfo']);
+  return PlatformAuthResult(user: user, additionalUserInfo: additionalUserInfo);
+}
+
+PlatformAdditionalUserInfo _decodeAdditionalUserInfo(
+    Map<dynamic, dynamic> data) {
+  return PlatformAdditionalUserInfo(
+    isNewUser: data['isNewUser'],
+    username: data['username'],
+    providerId: data['providerId'],
+    profile: data['profile']?.cast<String, dynamic>(),
+  );
+}
+
+PlatformIdTokenResult _decodeIdTokenResult(Map<String, dynamic> data) {
+  return PlatformIdTokenResult(
+    token: data['token'],
+    expirationTimestamp: data['expirationTimestamp'],
+    authTimestamp: data['authTimestamp'],
+    issuedAtTimestamp: data['issuedAtTimestamp'],
+    signInProvider: data['signInProvider'],
+    claims: data['claims'],
+  );
+}
+
+/// A utilily class that collects the callbacks for a [verifyPhoneNumber] call.
+class _PhoneAuthCallbacks {
+  _PhoneAuthCallbacks(
```

Comment:
Since everything is `final`, can the constructor be `const`?

Reply:
Yes.